[18.0][MIG] hr_attendance_report_theoretical_time: Migration to 18.0

xaviedoanhduy opened the pull request:
### Depends on: 

- [18.0][MIG] hr_holidays_public: Migration to 18.0: `https://github.com/OCA/hr-holidays/pull/147`
### OCA-Port

- [17.0][FW] hr_attendance_report_theoretical_time: port from 14.0: `https://github.com/OCA/hr-attendance/pull/181`

### On this occasion:
- Implementing `https://github.com/OCA/hr-holidays/pull/111`

pedrobaeza committed:
[ADD] hr_attendance_report_theoretical_time

Theoretical vs Attended Time Analysis
=====================================

This module adds a new report called "Theoretical vs Attended Time Analysis"
that compares worked time, measured through attendances records, with the
theoretical time, computed from employee's working calendar, public holidays
and employee specific leaves.

There is the possibility of counting as theoretical time some leave types if
specified in them.

Installation
============

On installation time, this module computes the theoretical hours for the day of
the attendance check-in, so if you have a lot of records, this would be a bit
slow.

Configuration
=============

You need to be at least "Attendance / Manual Attendance" for being able to see
the attendances report.

For including some leave types in the theoretical time, you have to:

* Go to Leaves > Configuration.
* Select leave type you want to include.
* Check the mark "Include in theoretical hours".

Usage
=====

* Go to *Attendances > Reports > Theoretical vs Attended Time Analysis*.
* Check pivot table or look at the graph view.

Known issues / Roadmap
======================

* Employees with less than 1 week in the company will show full week
  theoretical hours.
* Activate ORM cache for improving performance on computing theoretical hours,
  but assuring that the cache is cleared when the conditions of the computation
  changes.
* If you change employee's working time, theoretical hours for non attended
  days will be computed according this new calendar. You have to define
  start and end dates inside the calendar for avoiding this side effect.
* Recompute theoretical hours of affected days when changing the leave type
  to be included or not in theoretical time.

pedrobaeza committed:
[MIG] hr_attendance_report_theoretical_time: Migration to 12.0

* Standard procedure
* Adapt data model
* Adapt views
* Adapt code for date handling
* Adapt tests

[UPD] Update hr_attendance_report_theoretical_time.pot

Commit:
[12.0] hr_attendance_report_theoretical_time: add my attendance as default filter

Update translation files

Updated by "Update PO files to match POT (msgmerge)" hook in Weblate.

Translation: hr-12.0/hr-12.0-hr_attendance_report_theoretical_time
Translate-URL: https://translation.odoo-community.org/projects/hr-12-0/hr-12-0-hr_attendance_report_theoretical_time/

Commit:
[12.0][FIX] Travis Failing Holidays Leave Type Validity

Commit:
[12.0][IMP] hr_attendance_report_theoretical_time: Add filter wizard

[UPD] Update hr_attendance_report_theoretical_time.pot

hr_attendance_report_theoretical_time 12.0.1.1.0

Update translation files

Updated by "Update PO files to match POT (msgmerge)" hook in Weblate.

Translation: hr-12.0/hr-12.0-hr_attendance_report_theoretical_time
Translate-URL: https://translation.odoo-community.org/projects/hr-12-0/hr-12-0-hr_attendance_report_theoretical_time/

marylla committed:
Translated using Weblate (German)

Currently translated at 97.6% (41 of 42 strings)

Translation: hr-12.0/hr-12.0-hr_attendance_report_theoretical_time
Translate-URL: https://translation.odoo-community.org/projects/hr-12-0/hr-12-0-hr_attendance_report_theoretical_time/de/

chienandalu committed:
[IMP] hr_attendance_report_theoretical_time: recompute attendances

- We add a wizard so the already existing attendances can be recomputed
according to an employee calendar change, like a different for a set of
dates that weren't took into consideration in advance, so the wrong
theoretical hours are computed.

[UPD] Update hr_attendance_report_theoretical_time.pot

[UPD] README.rst

hr_attendance_report_theoretical_time 12.0.1.2.0

Update translation files

Updated by "Update PO files to match POT (msgmerge)" hook in Weblate.

Translation: hr-12.0/hr-12.0-hr_attendance_report_theoretical_time
Translate-URL: https://translation.odoo-community.org/projects/hr-12-0/hr-12-0-hr_attendance_report_theoretical_time/

chienandalu committed:
[12.0][FIX] hr_attendance_report_theoretical_time: menu permissions

- Although the minor permission to access the theoretical time report is
"Manual Attendance", the menu wasn't accessible due to permissions.
- Record rules also prevent such users from seeing other employees
attendances.

hr_attendance_report_theoretical_time 12.0.1.3.0

pegonzalezspesol committed:
[IMP] : black, isort

pegonzalezspesol committed:
[MIG] 13.0 hr_attendance_report_theoretical_time

MiquelRForgeFlow committed:
[FIX] Correct website in manifest

etobella committed:
[13.0][FIX] theoretical_time: Context was not imported on wizard

victoralmau committed:
[FIX] hr_attendance_report_theoretical_time: Show error in hr.lave form related to employee public (user without employee groups)

victoralmau committed:
[FIX] hr_attendance_report_theoretical_time: Remove search_default_my (context) when hr_attendance_theoretical_action is load from wizard

victoralmau committed:
[IMP] hr_attendance_report_theoretical_time: Add department_id in report

etobella committed:
[FIX] theoretical_time: Issue with postgres 12

ao-landoo committed:
[MIG] hr_attendance_report_theoretical_time: Migration to 14.0

[UPD] Update hr_attendance_report_theoretical_time.pot

[UPD] README.rst

ao-landoo committed:
[IMP] hr_attendance_report_theoretical_time: black, isort, prettier

victoralmau committed:
[FIX] hr_attendance_report_theoretical_time: Set default department_id in wizard correctly according to companies

victoralmau committed:
[FIX] hr_attendance_report_theoretical_time: Filter by all employees (including archived employees) in wizard

hr_attendance_report_theoretical_time 14.0.1.0.1

polimex-lgeorgiev committed:
Translated using Weblate (Bulgarian)

Currently translated at 96.1% (50 of 52 strings)

Translation: hr-attendance-14.0/hr-attendance-14.0-hr_attendance_report_theoretical_time
Translate-URL: https://translation.odoo-community.org/projects/hr-attendance-14-0/hr-attendance-14-0-hr_attendance_report_theoretical_time/bg/

yankinmax committed:
[FIX] hr_attendance_report_theoretical_time: access right to pivot report

hr_attendance_report_theoretical_time 14.0.1.0.2

victoralmau committed:
[IMP] hr_attendance_report_theoretical_time: Tests are added to validate the total hours in two-week calendars with defined start and end dates (solved since https://github.com/odoo/odoo/pull/96391).
TT34416

hr_attendance_report_theoretical_time 14.0.1.0.4

victoralmau committed:
[IMP] hr_attendance_report_theoretical_time: Separate the setup() of the tests in TestHrAttendanceReportTheoreticalTimeBase to improve the inheritance of tests in other modules.
TT34898

hr_attendance_report_theoretical_time 14.0.1.0.3

victoralmau committed:
[MIG] hr_attendance_report_theoretical_time: Migration to 15.0
TT36461

victoralmau committed:
[IMP] hr_attendance_report_theoretical_time: Reduce code according to https://github.com/odoo/odoo/pull/97507/commits/19f3a8d0e6f658ba05a5feee75aa53ae76ef52ad.
TT36461

chienandalu committed:
[FIX] hr_attendance_report_theoretical_time: tests

Tests rely on removing tz from the current user and it isn't possible
anymore after https://github.com/odoo/odoo/commit/d6f2437e9ce69a585c33940c6a2c18f4a6c7fdad#diff-38469def2f870bb866f971f57797dd7c21b6a95d52a8eae72f832f0eea2434f9R105

dzungtran89 committed:
[MIG] hr_attendance_report_theoretical_time: Migration to 16.0

victoralmau committed:
[IMP] hr_attendance_report_theoretical_time: Improve tests to better cover different use cases.

mymage committed:
Translated using Weblate (Italian)

Currently translated at 100.0% (53 of 53 strings)

Translation: hr-attendance-15.0/hr-attendance-15.0-hr_attendance_report_theoretical_time
Translate-URL: https://translation.odoo-community.org/projects/hr-attendance-15-0/hr-attendance-15-0-hr_attendance_report_theoretical_time/it/

Ivorra78 committed:
Translated using Weblate (Spanish)

Currently translated at 100.0% (53 of 53 strings)

Translation: hr-attendance-16.0/hr-attendance-16.0-hr_attendance_report_theoretical_time
Translate-URL: https://translation.odoo-community.org/projects/hr-attendance-16-0/hr-attendance-16-0-hr_attendance_report_theoretical_time/es/

victoralmau committed:
[FIX] hr_attendance_report_theoretical_time: Show the Reports menu to a basic user

We need to leave the action empty for consistency because we are going
to set different submenus, otherwise the Reports menu would not be
displayed to a basic user.

TT44849

nilshamerlinck committed:
Translated using Weblate (French)

Currently translated at 47.1% (25 of 53 strings)

Translation: hr-attendance-16.0/hr-attendance-16.0-hr_attendance_report_theoretical_time
Translate-URL: https://translation.odoo-community.org/projects/hr-attendance-16-0/hr-attendance-16-0-hr_attendance_report_theoretical_time/fr/

juanjosesegui-tecnativa committed:
[IMP] hr_attendance_report_theoretical_time: pre-commit auto fixes

juanjosesegui-tecnativa committed:
[MIG] hr_attendance_report_theoretical_time: Migration to 17.0

weblate committed:
Update translation files

Updated by "Update PO files to match POT (msgmerge)" hook in Weblate.

Translation: hr-attendance-17.0/hr-attendance-17.0-hr_attendance_report_theoretical_time
Translate-URL: https://translation.odoo-community.org/projects/hr-attendance-17-0/hr-attendance-17-0-hr_attendance_report_theoretical_time/

xaviedoanhduy committed:
[MIG] hr_attendance_report_theoretical_time: Migration to 18.0

xaviedoanhduy committed:
[DON'T MERGE] Add test-requirements.txt

chaule97 commented on hr_attendance_report_theoretical_time/tests/test_hr_attendance_report_theoretical_time.py:
```diff
@@ -99,11 +99,16 @@ def setUpClass(cls):
+            UPDATE hr_employee
+            SET create_date = '{}'
+            WHERE id IN ({}, {})
+        """.format("1946-12-23 12:00:00", cls.employee_1.id, cls.employee_2.id)
```
```suggestion
       from odoo.tools import SQL

       cls.env.cr.execute(
            SQL(
                "UPDATE hr_employee SET create_date = %s WHERE id in (%s, %s)",
                "1946-12-23 12:00:00",
                cls.employee_1.id,
                cls.employee_2.id,
            )
        )
```

you can use new [SQL wrapper](https://www.odoo.com/documentation/18.0/developer/reference/backend/orm.html#odoo.tools.SQL) feature

chaule97 commented on hr_attendance_report_theoretical_time/wizards/wizard_theoretical_time.py:
```diff
@@ -0,0 +1,63 @@
+        domain = self._prepare_employee_domain()
+        self.employee_ids = self.env["hr.employee"].search(domain)
+        action = {
+            "name": _("Select Employees to Analyze Theoretical Time"),
```
```suggestion
            "name": self.env._("Select Employees to Analyze Theoretical Time"),
```

chaule97 commented on hr_attendance_report_theoretical_time/wizards/wizard_theoretical_time.py:
```diff
@@ -0,0 +1,63 @@
+# Copyright 2021 Tecnativa - Víctor Martínez
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+
+from odoo import _, api, fields, models
```
```suggestion
from odoo import api, fields, models
```

xaviedoanhduy commented:
thanks @chaule97, the PR has been updated

chaule97 reviewed (approved):
LGTM

OCA-git-bot committed:
[IMP] update dotfiles

victoralmau committed:
[IMP] hr_attendance_report_theoretical_time: Add company compatibility in the report

If only one company is selected, only the data of the employees of that company should be displayed.

TT51401

victoralmau committed:
[IMP] hr_attendance_report_theoretical_time: Add Active employees filter to report

TT51401
